File: backend/src/modules/catalog/services/otp.service.ts
import { OTPModel } from '../models/otp.model';
import crypto from 'crypto';
import { hashPassword, comparePassword } from '../../../common/utils/password';
import { AppError } from '../../../common/errors/app-error';
import twilio from 'twilio';
import { EmailService } from './email.service';

const accountSid = process.env.TWILIO_ACCOUNT_SID;
const authToken = process.env.TWILIO_AUTH_TOKEN;
const verifyServiceSid = process.env.TWILIO_VERIFY_SERVICE_SID;

const client = accountSid && authToken ? twilio(accountSid, authToken) : null;

export class OTPService {
    private static ipRequestMap = new Map<string, { count: number; lastReset: number }>();

    static async checkThrottling(ip: string) {
        const now = Date.now();
        const record = this.ipRequestMap.get(ip) || { count: 0, lastReset: now };

        if (now - record.lastReset > 60000) { // Reset every minute
            record.count = 0;
            record.lastReset = now;
        }

        // In development/load testing, allow much higher limits
        const limit = process.env.NODE_ENV === 'development' ? 10000 : 100;

        if (record.count >= limit) { // Max requests per minute per IP
            throw new AppError('Too many requests. Please try again in a minute.', 429);
        }

        record.count += 1;
        this.ipRequestMap.set(ip, record);
    }

    static async sendSMS(phoneNumber: string, otp: string): Promise<boolean> {
        if (!client || !verifyServiceSid) {
            console.warn(`[SMS Mock] TWILIO credentials missing. Logged OTP: ${otp} to ${phoneNumber}`);
            return true;
        }

        const fromNumber = process.env.TWILIO_PHONE_NUMBER;

        if (!fromNumber) {
            console.error('[SMS Error] TWILIO_PHONE_NUMBER is missing in .env');
            return false;
        }

        try {
            // Use Programmable SMS to send the EXACT OTP we generated
            await client.messages.create({
                body: `Your The Souled Store verification code is: ${otp}. Valid for 5 minutes.`,
                from: fromNumber,
                to: phoneNumber
            });

            console.log(`[Twilio SMS] OTP sent to ${phoneNumber}`);
            return true;
        } catch (error: any) {
            console.error('[Twilio Error]', error.message);
            return false;
        }
    }

    static async sendEmail(email: string, otp: string): Promise<boolean> {
        return await EmailService.sendOTP(email, otp);
    }

    /**
     * Sends OTP to both Phone and Email
     */
    static async sendDualOTP(phoneNumber: string | undefined, email: string, otp: string): Promise<boolean> {
        console.log('[OTP] Sending to email:', email);
        const emailSent = await this.sendEmail(email, otp);
        console.log('[OTP] Email sent:', emailSent);

        let smsSent = true;

        if (phoneNumber) {
            console.log('[OTP] Sending to phone:', phoneNumber);
            smsSent = await this.sendSMS(phoneNumber, otp);
            console.log('[OTP] SMS sent:', smsSent);
        }

        // Return true if at least one succeeded, or as per preference. 
        // User said "implement real time otp for mobile and mail", implying both.
        const result = emailSent || smsSent;
        console.log('[OTP] Final result:', result);
        return result;
    }

    static async generateOTP(userId: string, type: 'registration' | 'password_reset' | 'login_verification' | 'phone_verification') {
        const otp = crypto.randomInt(100000, 999999).toString();
        const expiresAt = new Date(Date.now() + 5 * 60 * 1000); // 5 minutes

        const otpHash = await hashPassword(otp);

        await OTPModel.deleteMany({ userId, type });
        await OTPModel.create({
            userId,
            otpHash,
            type,
            expiresAt,
        });

        return otp;
    }

    static async verifyOTP(userId: string, otp: string, type: 'registration' | 'password_reset' | 'login_verification' | 'phone_verification', persist: boolean = false) {
        const otpDoc = await OTPModel.findOne({ userId, type });

        if (!otpDoc) {
            throw new AppError('OTP expired or not found', 400);
        }

        if (otpDoc.attempts >= 3) {
            await OTPModel.findByIdAndDelete(otpDoc._id);
            throw new AppError('Too many failed attempts. Please request a new code.', 400);
        }

        const isValid = await comparePassword(otp, otpDoc.otpHash);

        if (!isValid) {
            otpDoc.attempts += 1;
            await otpDoc.save();
            throw new AppError('Invalid verification code', 400);
        }

        // Only delete if NOT persisting (persistence is needed for two-step flows like password reset)
        if (!persist) {
            await OTPModel.findByIdAndDelete(otpDoc._id);
        }
        return true;
    }
}

File: backend-spring/src/main/java/com/threads/catalog/service/OtpService.java
package com.threads.catalog.service;

import com.threads.catalog.model.Otp;
import com.threads.catalog.repository.OtpRepository;
import com.threads.common.exception.ResourceNotFoundException;
import com.twilio.Twilio;
import com.twilio.rest.api.v2010.account.Message;
import com.twilio.type.PhoneNumber;
import jakarta.annotation.PostConstruct;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.security.SecureRandom;
import java.util.Date;
import java.util.Optional;

@Service
@Slf4j
@RequiredArgsConstructor
public class OtpService {

    private final OtpRepository otpRepository;
    private final EmailService emailService;
    private final PasswordEncoder passwordEncoder;

    @Value("${twilio.account-sid}")
    private String twilioAccountSid;

    @Value("${twilio.auth-token}")
    private String twilioAuthToken;

    @Value("${twilio.from-number}")
    private String twilioFromNumber;

    @PostConstruct
    public void init() {
        log.info("OtpService: Twilio Account SID = {}", (twilioAccountSid != null && !twilioAccountSid.isEmpty()) ? "PRESENT" : "MISSING");
        if (twilioAccountSid != null && !twilioAccountSid.isEmpty() && twilioAuthToken != null && !twilioAuthToken.isEmpty()) {
            Twilio.init(twilioAccountSid, twilioAuthToken);
            log.info("Twilio client initialized");
        } else {
            log.warn("Twilio credentials not configured. SMS will be mocked.");
        }
    }

    public boolean sendSMS(String phoneNumber, String otpCode) {
        if (twilioAccountSid == null || twilioAccountSid.isEmpty()) {
            String maskedPhone = phoneNumber != null && phoneNumber.length() > 4 ? phoneNumber.substring(0, 2) + "****" + phoneNumber.substring(phoneNumber.length() - 2) : "****";
            log.warn("[SMS Mock] TWILIO credentials missing. OTP handled for: {}", maskedPhone);
            return true;
        }
        try {
            Message message = Message.creator(
                    new PhoneNumber(phoneNumber),
                    new PhoneNumber(twilioFromNumber),
                    "Your Threads Fashion verification code is: " + otpCode + ". Valid for 5 minutes."
            ).create();
            log.info("[Twilio SMS] OTP sent to {}", phoneNumber);
            return true;
        } catch (Exception e) {
            log.error("[Twilio Error] {}", e.getMessage());
            return false;
        }
    }

    public boolean sendDualOTP(String phoneNumber, String email, String otpCode) {
        log.info("[OTP] Sending to email: {}", email);
        emailService.sendOTP(email, otpCode);
        boolean emailSent = true; // Email service sends async currently, assuming success or mocked
        
        boolean smsSent = false;
        if (phoneNumber != null && !phoneNumber.isEmpty()) {
            log.info("[OTP] Sending to phone: {}", phoneNumber);
            smsSent = sendSMS(phoneNumber, otpCode);
        }
        
        return emailSent || smsSent;
    }

    public String generateOTP(String userId, String type) {
        SecureRandom random = new SecureRandom();
        int num = 100000 + random.nextInt(900000);
        String otpCode = String.valueOf(num);

        String otpHash = passwordEncoder.encode(otpCode);
        
        otpRepository.deleteByUserIdAndType(userId, type);
        
        Otp otp = Otp.builder()
                .userId(userId)
                .otpHash(otpHash)
                .type(type)
                .expiresAt(new Date(System.currentTimeMillis() + 5 * 60 * 1000)) // 5 mins
                .attempts(0)
                .build();
                
        otpRepository.save(otp);
        return otpCode;
    }

    public boolean verifyOTP(String userId, String otpCode, String type, boolean persist) {
        Optional<Otp> otpDocOpt = otpRepository.findByUserIdAndType(userId, type);
        
        if (otpDocOpt.isEmpty()) {
            throw new ResourceNotFoundException("OTP expired or not found");
        }
        
        Otp otpDoc = otpDocOpt.get();
        
        if (otpDoc.getAttempts() >= 3) {
            otpRepository.deleteById(otpDoc.getId());
            throw new ResourceNotFoundException("Too many failed attempts. Please request a new code.");
        }
        
        if (!passwordEncoder.matches(otpCode, otpDoc.getOtpHash())) {
            otpDoc.setAttempts(otpDoc.getAttempts() + 1);
            otpRepository.save(otpDoc);
            throw new ResourceNotFoundException("Invalid verification code");
        }
        
        if (!persist) {
            otpRepository.deleteById(otpDoc.getId());
        }
        return true;
    }
}
